File: 11_container_with_most_water.go
/*
 * @lc app=leetcode id=11 lang=golang
 *
 * [11] Container With Most Water
 */

// @lc code=start
func maxArea(height []int) int {
	left, right := 0, len(height)-1
	m := 0
	for left < right {
		m = max(m, min(height[left], height[right])*long(left, right))
		if height[left] < height[right] {
			left++
			continue
		}
		right--
	}
	return m
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func long(a, b int) int {
	return b - a
}

// @lc code=end

File: 11_container_with_most_water.rs
/*
 * @lc app=leetcode id=11 lang=rust
 *
 * [11] Container With Most Water
 */

// @lc code=start
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let (mut left, mut right, mut ans) = (0 as usize, height.len() - 1, -1);
        while left < right {
            ans = ans.max(height[left].min(height[right]) * (right as i32 - left as i32));
            if height[left] < height[right] {
                left += 1;
            } else {
                right -= 1;
            }
        }
        ans
    }
}

// @lc code=end

